google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/ReadOnlyStalenessUtil.java

```diff
@@ -0,0 +1,260 @@
+  /**
+   * Parses an RFC3339 date/time value with millisecond precision and returns this as a {@link
+   * Timestamp}. Although {@link Timestamp} has nanosecond precision, this method is only used for
+   * parsing read-only staleness values, and more than millisecond precision is not needed for that.
+   */
+  static Timestamp parseRfc3339(String str) throws SpannerException {
+    try {
+      DateTime dateTime = DateTime.parseRfc3339(str);
+      return Timestamp.ofTimeMicroseconds(dateTime.getValue() * 1000L);
+    } catch (NumberFormatException e) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.INVALID_ARGUMENT, String.format("Invalid timestamp: %s", str), e);
+    }
+  }
+
+  /** The abbreviations for time units that may be used for client side statements. */
+  enum TimeUnitAbbreviation {
+    NANOSECONDS("ns", TimeUnit.NANOSECONDS),
+    MICROSECONDS("us", TimeUnit.MICROSECONDS),
+    MILLISECONDS("ms", TimeUnit.MILLISECONDS),
+    SECONDS("s", TimeUnit.SECONDS);
+
+    private final String abbreviation;
+    private final TimeUnit unit;
+
+    private TimeUnitAbbreviation(String abbreviation, TimeUnit unit) {
+      this.abbreviation = abbreviation;
+      this.unit = unit;
+    }
+
+    String getAbbreviation() {
+      return abbreviation;
+    }
+
+    TimeUnit getUnit() {
+      return unit;
+    }
+  }
+
+  /** Get the abbreviation for the given {@link TimeUnit}. */
+  static String getTimeUnitAbbreviation(TimeUnit unit) {
+    for (TimeUnitAbbreviation abb : TimeUnitAbbreviation.values()) {
+      if (abb.unit == unit) return abb.abbreviation;
+    }
+    throw SpannerExceptionFactory.newSpannerException(
+        ErrorCode.INVALID_ARGUMENT, "Invalid option for time unit: " + unit);
+  }
+
+  /** Get the {@link TimeUnit} corresponding with the given abbreviation. */
+  static TimeUnit parseTimeUnit(String unit) {
+    for (TimeUnitAbbreviation abb : TimeUnitAbbreviation.values()) {
+      if (abb.abbreviation.equalsIgnoreCase(unit)) return abb.unit;
+    }
+    throw SpannerExceptionFactory.newSpannerException(
+        ErrorCode.INVALID_ARGUMENT, "Invalid option for time unit: " + unit);
+  }
+
+  /**
+   * Internal interface that is used to generalize getting a time duration from Cloud Spanner
+   * read-only staleness settings.
+   */
+  static interface DurationValueGetter {
+    long getDuration(TimeUnit unit);
+
+    boolean hasDuration();
+  }
+
+  static final class GetExactStaleness implements DurationValueGetter {
+    private final TimestampBound staleness;
+
+    public GetExactStaleness(TimestampBound staleness) {
+      this.staleness = staleness;
+    }
+
+    @Override
+    public long getDuration(TimeUnit unit) {
+      return staleness.getExactStaleness(unit);
+    }
+
+    @Override
+    public boolean hasDuration() {
+      return staleness.getMode() == Mode.EXACT_STALENESS;
+    }
+  }
+
+  static final class GetMaxStaleness implements DurationValueGetter {
```

Comment:
minor - can address now or in future commit - I would rename to "MaxStalenessGetter"